cmake/modules/autopas_CodeCoverage.cmake

```diff
@@ -1,19 +1,44 @@
-option(AUTOPAS_CODE_COVERAGE "Code Coverage" OFF)
+set(LCOV_ENABLED_MSG "LCOV Coverage report enabled")
+set(LCOV_DISABLED_MSG "LCOV Coverage report disabled")
+set(DISABLE_COVERAGE_MSG "Disabling code coverage. CMAKE_BUILD_TYPE=Debug and AUTOPAS_BUILD_TESTS=ON required.")
+set(COVERAGE_REQUIREMENTS_MSG "Enable code coverage, requires GCC and CMAKE_BUILD_TYPE=Debug.")
+set(NON_GNU_COMPILER_ERROR_MSG "For code coverage ensure that GCOV, LCOV and GENHTML is installed. Disabling code coverage.")
 
-if (AUTOPAS_CODE_COVERAGE MATCHES ON)
-    if (CMAKE_BUILD_TYPE MATCHES Debug)
-        message(STATUS "ENABLING CODE COVERAGE TARGET")
-        include(CodeCoverage)
-        set(CMAKE_CXX_FLAGS_DEBUG "${CMAKE_CXX_FLAGS_COVERAGE} ${CMAKE_CXX_FLAGS_DEBUG}")
-        setup_target_for_coverage(${PROJECT_NAME}_coverage runTests coverage)
-        setup_target_for_coverage_cobertura(
-            ${PROJECT_NAME}_cobertura
-            runTests
-            coverage
-            --gtest_output=xml:coverage.junit.xml
-            .*gtest.*
-        )
-    else ()
-        message(FATAL_ERROR "Code Coverage only works with CMAKE_BUILD_TYPE = Debug")
-    endif ()
-endif ()
+option(AUTOPAS_ENABLE_COVERAGE ${COVERAGE_REQUIREMENTS_MSG} OFF)
+
+if (AUTOPAS_ENABLE_COVERAGE)
+
+  find_program(GCOV_EXECUTABLE gcov)
+  find_program(LCOV_EXECUTABLE lcov)
+  find_program(GENHTML_EXECUTABLE genhtml)
+
+  if (NOT GCOV_EXECUTABLE OR NOT LCOV_EXECUTABLE OR NOT GENHTML_EXECUTABLE)
+    message(FATAL_ERROR ${COVERAGE_REQUIREMENTS_MSG})
+  endif ()
+
+  message(STATUS ${LCOV_ENABLED_MSG})
+  if (NOT AUTOPAS_BUILD_TESTS OR NOT CMAKE_BUILD_TYPE STREQUAL "Debug")
+    message(WARNING ${DISABLE_COVERAGE_MSG})
+    set(AUTOPAS_ENABLE_COVERAGE OFF CACHE BOOL ${COVERAGE_REQUIREMENTS_MSG} FORCE)
+  endif()
+  if (NOT CMAKE_CXX_COMPILER_ID STREQUAL "GNU")
+    message(FATAL_ERROR ${NON_GNU_COMPILER_ERROR_MSG})
+    set(AUTOPAS_ENABLE_COVERAGE OFF CACHE BOOL ${COVERAGE_REQUIREMENTS_MSG} FORCE)
+  endif()
+
+  if(TARGET runTests)
+    target_compile_options(runTests PRIVATE --coverage)
+  endif()
+
+  # add coverage target
+  add_custom_target(
+    coverage
+    DEPENDS runTests
+    COMMAND ${CMAKE_BINARY_DIR}/tests/testAutopas/runTests
+    COMMAND ${LCOV_EXECUTABLE} --ignore-errors mismatch --rc branch_coverage=1 --directory . --capture --exclude */tests/* --exclude */build/* --exclude '/usr/*' --output-file coverage.info
```

Comment:
When I run `make coverage` I get the following error (last line) after all tests have run and no `coverage/` folder is created. 

```bash
Capturing coverage data from .
Subroutine read_intermediate_text redefined at /usr/bin/geninfo line 2623.
Subroutine read_intermediate_json redefined at /usr/bin/geninfo line 2655.
Subroutine intermediate_text_to_info redefined at /usr/bin/geninfo line 2703.
Subroutine intermediate_json_to_info redefined at /usr/bin/geninfo line 2792.
Subroutine get_output_fd redefined at /usr/bin/geninfo line 2872.
Subroutine print_gcov_warnings redefined at /usr/bin/geninfo line 2900.
Subroutine process_intermediate redefined at /usr/bin/geninfo line 2930.
Found gcov version: 11.3.0
Using intermediate gcov format
geninfo: ERROR: unknown argument for --ignore-errors: mismatch
```

I guess the reason is this line? Why do we need this anyway? Is it maybe from a newer / older version of `lcov`?

Ubuntu 22.04
```bash
$ lcov --version
lcov: LCOV version 1.14
```

Reply:
For the `mismatch` argument for `--ignore-errors` we need a more recent LCOV version. I have 2.1 installed here.
There seem to be some inconsistencies in the coverage output after running the tests. The output message suggested to use `--ignore-errors: mismatch` to ignore them.

Reply:
Digging through the [man page of the tagged releases](https://github.com/linux-test-project/lcov/blob/master/man/lcov.1) this feature seems to exist since `lcov 2.0`. I'm fine with that but we should state this required version in the documentation and if possible check it with CMake.

Reply:
I have added a check for the LCOV version in [f69aca2](https://github.com/AutoPas/AutoPas/pull/888/commits/f69aca2f80e7dc144a25dc24fa94d64337ad9258). This is tested with lcov 1.14 -> error and lcov 2.1 -> success.
Testing.md now states that lcov version >= 2.0 is a requirement.